Software/Development/SemesterProject/App/referenceHandler.c

```diff
@@ -0,0 +1,317 @@
+/*
+ * referenceHandler.c
+ *
+ *  Created on: Mar 27, 2019
+ *      Author: Nicolas
+ *
+ *      >>Looking at the system description diagram found in the class design
+ *      is highly recommended<<
+ *
+ *      The goal of this module is to check what reference source is selected
+ *      by the user. When the source is known, the value of that reference must
+ *      be obtained. When the value is obtained and with the goal of limiting
+ *      large reference differences a new reference considering the maximum
+ *      allowed reference derivative is calculated. This is calculated by setting
+ *      a maximum deltaTorque and deltaSpeed.
+ *
+ *      For example, if the reference is changed from 10 Nm to 20 Nm, it will
+ *      increase deltaTorque every time the referenceHandler is run.
+ *      If deltaTorque is 0.1 and the referenceHandler is run every 20 ms.
+ *      The new reference will be reached in (20 - 10)/(0.1/20e-3) = 2 s. The
+ *      key parameter is deltaTorque and the user can select the desired value.
+ *
+ *      Then, in order to avoid references over the technical limits of the system,
+ *      like a 100 Nm torque reference, the references are limited to a preset
+ *      value MAX_TORQUE_REF_NM and MAXIMUM_SPEED_REF_RPM.
+ *
+ *      The interfaces provided are:
+ *          *Torque reference
+ *          *Speed reference
+ *          *Open-loop reference (which collects the reference from torque inputs)
+ *          *Cruise control or Torque control
+ *          *Reference source
+ *          *Reference source change
+ *
+ *      The scheduled function is handleReferences().
+ *      In this module there are two tasks performed:
+ *      ** Reference calculation    calculateReference()
+ *      ** Reference type (torque/cruise control) decision  decideReferenceType()
+ *
+ *
+ */
+
+#include "referenceHandler.h"
+
+/*
+ * Quasi-global definition
+ */
+static GUISignalsTag GUISignals;
+static float speedReference, torqueReference;
+static float deltaSpeed = 5, deltaTorque = 0.1;
+static int undampedSpeedReference = 0;
+static int referenceSourceChanged = 0;
+
+
+static referenceSourceTag referenceSource = interfacePCB;
+static enum referenceTypeTag referenceType = torqueControl;
+const int SPEED_STEP_FROM_BUTTON_RPM = 60 * SPEED_STEP_FROM_BUTTON_KPH / 3.6 * GEAR_RATIO / RADIUS_WHEELS;
+
+
+void handleReferences(void)
+{
+    GUISignals = getGUISignals();
+    decideReferenceSource();
```

Comment:
Because the GUISignals include the Reference Source toggle, then in the case that the system is not running, you wouldn't be able to change the reference source.